[Clang][OpenMP] Capture mapped pointers on `target` by reference.

abhinavgaba opened the pull request:
For the following:

```c
int *p;
  #pragma omp target map(p[0]) //    (A)
  (void)p;

  #pragma omp target map(p) //       (B)
  (void)p;

  #pragma omp target map(p, p[0]) // (C)
  (void)p;

  #pragma omp target map(p[0], p) // (D)
  (void)p;
```

For (A), the pointer `p` is predetermined `firstprivate`, so it should be (and is) captured by-copy. However, for (B), (C), and (D), since `p` is already listed in a `map` clause, it's not predetermined `firstprivate`, and hence, should be captured by-reference, like any other mapped variable.

To ensure the correct handling of (C) and (D), the following changes were made:

1. In SemaOpenMP, we now ensure that `p` is marked to be captured by-reference in these cases.

2. We no longer ignore `map(p)` during codegen of `target` constructs, even if there's another map like `map(p[0])` that would have been mapped using a PTR_AND_OBJ map.

3. For cases like (D), we now handle `map(p)` before `map(p[0])`, so the former gets the TARGET_PARAM flag and sets the kernel argument.

abhinavgaba committed:
[Clang][OpenMP] Capture mapped pointers on `target` by reference.

For the following:

```c
int *p;
  \#pragma omp target map(p[0]) //    (A)
  (void)p;

  \#pragma omp target map(p) //       (B)
  (void)p;

  \#pragma omp target map(p, p[0]) // (C)
  (void)p;

  \#pragma omp target map(p[0], p) // (D)
  (void)p;
```

For (A), the pointer `p` is predetermined `firstprivate`, so it
should be (and is) captured by-copy. However, for (B), (C), and
(D), since `p` is already listed in a `map` clause, it's not
predetermined `firstprivate`, and hence, should be captured
by-reference, like any other mapped variable.

To ensure the correct handling of (C) and (D), the following
changes were made:

1. In SemaOpenMP, we now ensure that `p` is marked to be
captured by-reference in these cases.

2. We no longer ignore `map(p)` during codegen of `target`
constructs, even if there's another map like `map(p[0])` that
would have been mapped using a PTR_AND_OBJ map.

3. For cases like (D), we now handle `map(p)` before `map(p[0])`,
so the former gets the TARGET_PARAM flag and sets the kernel
argument.

abhinavgaba committed:
Add another test, minor clang-format fix.

abhinavgaba commented:
@alexey-bataev, please review.

abhinavgaba commented on clang/lib/CodeGen/CGOpenMPRuntime.cpp:
```diff
@@ -8811,8 +8829,19 @@ class MappableExprsHandler {
-                                             const MapData &RHS) {
+    llvm::stable_sort(DeclComponentLists, [VD](const MapData &LHS,
+                                               const MapData &RHS) {
+      // For cases like map(p, p[0], p[0][0]), the shortest map, like map(p)
+      // in this case, should be handled first, to ensure that it gets the
+      // TARGET_PARAM flag.
+      OMPClauseMappableExprCommon::MappableExprComponentListRef Components =
+          std::get<0>(LHS);
+      OMPClauseMappableExprCommon::MappableExprComponentListRef ComponentsR =
+          std::get<0>(RHS);
+      if (VD && VD->getType()->isAnyPointerType() && Components.size() == 1 &&
+          ComponentsR.size() > 1)
+        return true;
```
Is there a case you have in mind? If the number of components is same, then this should not return anything, and we continue to the existing logic and return based on that in line 8855.

The new return should only kick-in if the current component list has one entry and the other one has more.

I originally wanted to use the following, which is broader, (and the comment still is in line with this):

```cpp
if (VD && VD->getType()->isAnyPointerType() && Components.size() !=
          ComponentsR.size())
  return Components.size() < ComponentsR.size();
```

because even when we have `map(p[0][0], p[0])`, the `map(p[0])` should contribute to the kernel argument. `p` is predetermined `firstprivate because `p` is the base-pointer of `map(p[0])`. 

The base-pointer for `int **p; ... map(p[0][0])` is `p[0]`, so it has no effect on the data-sharing/mapping of `p`, and hence should not determine the kernel argument.

But technically we don't need to sort between `p[0][0]` and `p[0][0][0]` since neither of them should ideally contribute to the determination of the data-sharing/mapping clause on `p`.

Note that sorting between `p[0]` and `p[0][0]` is not needed _yet_, since both `p[0][0]` and `p[0]` use PTR_AND_OBJ mappings that start with `p`.

----
Future Plans:

I am working on a change to support OpenMP 5.0/6.1 compliant pointer-attachment that avoids using the PTR_AND_OBJ mapping for them, and instead uses a new ATTACH map-type bit. With that, the map-chain emitted for `map(present, to: p[0][0]) map(from: p[0])` would look like the following:

```c
// For p[0][0]
&p[0][0], &p[0][0], sizeof(p[0][0]), TO | PRESENT
&p[0], &p[0][0], sizeof(ptr), ATTACH

// For p[0]
&p[0], &p[0], sizeof(p[0]), FROM | PARAM
&p, &p[0], ATTACH

ATTACH:
* New map-type bit which means don't increase the ref-count, only do an attachment iff either `&p`'s ref-count is 1, or `&p[0]`'s ref-count is 1 after finishing all maps, including mappers.
* Can be combined with ALWAYS for `attach(always)` map-type-modifier to force the attachment even if the ref-counts are more than 1.
* Can be removed if the user says `attach(never)` (OpenMP 6.1)
```

Once the above is supported, we would need the PARAM bit to apply to the map of `p[0]`, which would likely require sorting between the component-lists, even if neither is of size 1.

abhinavgaba committed:
Update the comparator for sort, plus minor additions to offload tests.

abhinavgaba commented on clang/lib/CodeGen/CGOpenMPRuntime.cpp:
```diff
@@ -8811,8 +8829,19 @@ class MappableExprsHandler {
-                                             const MapData &RHS) {
+    llvm::stable_sort(DeclComponentLists, [VD](const MapData &LHS,
+                                               const MapData &RHS) {
+      // For cases like map(p, p[0], p[0][0]), the shortest map, like map(p)
```
And I'm also not sure yet the lambda captures function would interact with this if we split g`enerateInfoForCapture.

> 2. What if we have 3-level dependency, will we need the third loop in this case?

For a three-level pointer case like this:

`map(to: sp->x, sp->y) map(to: sp->sq->a, sp->sq->b) map(to: sp->sq->sr->c, sp->sq->sr->d) map(sp)`

We should eventually have something like:
```
&sp, &sp, sizeof(sp), TO | PARAM

// map-chain for the containing-struct sp[0] with base-pointer sp
&sp[0], &sp[0], sizeof(sp[0]), ALLOC
&sp[0], &sp->x, sizeof(sp->x), TO | MEMBER_OF(2)
&sp[0], &sp->y, sizeof(sp->y), TO | MEMBER_OF(2)
&sp, &sp[0], ATTACH

// map-chain for the containing-struct sp->sq[0] with base-pointer sp->sq
&(sp->sq[0]), &(sp->sq[0]), sizeof(sp->sq[0]), ALLOC
&sp->sq[0], &sp->sq->a, sizeof(sp->sq->a), TO | MEMBER_OF(6)
&sp->sq[0], &sp->sq->a, sizeof(sp->sq->a), TO | MEMBER_OF(6)
&sp->sq, &sp->sq[0], ATTACH

// map-chain for the containing-struct sp->sq->sr[0] with base-pointer sp->sq->sr
&(sp->sq->sr[0]), &(sp->sq->sr[0]), sizeof(sp->sq->sr[0]), ALLOC
&sp->sq->sr[0], &sp->sq->sr->c, sizeof(sp->sq->sr->c), TO | MEMBER_OF(10)
&sp->sq->sr[0], &sp->sq->sr->d, sizeof(sp->sq->sr->d), TO | MEMBER_OF(10)
&sp->sq->sr, &sp->sq->sr[0], ATTACH
```

The `for (Expr* AttachBasePtr: AttachBasePtrs)` loop would need to run three times, once each for the for the attachable-base-pointer `sp`, `sp->sq`, and `sp->sq->sr`.

And we can think of the first invocation to be for the component-lists that have no base-pointer, and pull that into the loop as well.

alexey-bataev replied:
Try to fix in a single loop at first. If it does not work, try splitting

abhinavgaba replied:
I created two NFC PRs: https://github.com/llvm/llvm-project/pull/146934 and https://github.com/llvm/llvm-project/pull/146891 to add new tests and do a refactor of this code a bit. Please take a look. I have a patch locally for this PR that builds upon the refactor. I can upload that once those NFC PRs are merged.

It would fix the IR emission for this case, but there would still be an underlying problem for the struct member case -- since we don't use PTR_AND_OBJ for mapping things like "sp->x", when we have "map(sp, sp->x)", we can either access the mapped pointer, or the pointee inside the target region, because sp and sp->x are not attached. So if we merge the fix for mapped "sp" not being accessible in the target region, it will make the mapped pointee inaccessible. One of the tests from the NFC PR shows this problem, and I can highlight that once those are merged and I upload a follow-up patch to this one.

abhinavgaba committed:
[NFC][Clang] Refactor mapinfo generation for captured vars on target.

abhinavgaba committed:
[NFC][Clang] Refactor mapinfo generation for captured vars on target.

abhinavgaba committed:
Handle mapping of captured pointers before any dereference of them.

abhinavgaba committed:
No need to change the sorting since we now handle mapping of pointers before their derefs.

abhinavgaba committed:
Minor clang-format fix.

abhinavgaba committed:
[NFC][OpenMP] Add tests for mapping pointers and their dereferences.

The output of the compile-and-run tests is incorrect. These will be
used for reference in future commits that resolve the issues.

Also updated the existing clang LIT test,
target_map_both_pointer_pointee_codegen.cpp, with more regions and
more narrowed-down update_cc_test_checks filters.

abhinavgaba committed:
Minor clang-format fix.

abhinavgaba committed:
Merge branch 'capture-info-refactor' into tgt-capture-mapped-ptrs-by-ref

abhinavgaba committed:
Merge branch 'nfc-lit-test-updates-for-incoming-pr' into tgt-capture-mapped-ptrs-by-ref

abhinavgaba committed:
Minor NFC update to a test.

abhinavgaba committed:
Update tests with the current output.

abhinavgaba committed:
Merge remote-tracking branch 'origin/main' into tgt-capture-mapped-ptrs-by-ref

alexey-bataev commented:
> For the following:
> 
> ```c
> int *p;
>   #pragma omp target map(p[0]) //    (A)
>   (void)p;
> 
>   #pragma omp target map(p) //       (B)
>   (void)p;
> 
>   #pragma omp target map(p, p[0]) // (C)
>   (void)p;
> 
>   #pragma omp target map(p[0], p) // (D)
>   (void)p;
> ```
> 
> For (A), the pointer `p` is predetermined `firstprivate`, so it should be (and is) captured by-copy. However, for (B), (C), and (D), since `p` is already listed in a `map` clause, it's not predetermined `firstprivate`, and hence, should be captured by-reference, like any other mapped variable.
> 
> To ensure the correct handling of (C) and (D), the following changes were made:
> 
> 1. In SemaOpenMP, we now ensure that `p` is marked to be captured by-reference in these cases.
> 2. We no longer ignore `map(p)` during codegen of `target` constructs, even if there's another map like `map(p[0])` that would have been mapped using a PTR_AND_OBJ map.
> 3. For cases like (D), we now handle `map(p)` before `map(p[0])`, so the former gets the TARGET_PARAM flag and sets the kernel argument.

I have one big question - why do we need this? If pointer is mapped explicitly, it becomes an attached pointer, which cannot be modified in target region. How we capture/process it, byref or byval, does not matter, we can do whatever is better. So, why do we need to process it byref?

abhinavgaba commented:
> Even for cases where both the pointer, pointee are mapped, the expectation is for the pointer to still be captured by-ref.
>Why?

Because we don't know if the pointer/pointee were both already existing on the device, in which case, there should be no pointer-attachment for them.

```c
  int x = 111;
  #pragma omp target enter data map(x) // map the pointee
  int *p = &x;
  #pragma omp target enter data map(p) // map the pointer

  #pragma omp target map(p, p[0:1]) // No pointer-attachment between p and p[0], since both
                                    // already existed on the device
  {
    p = (int*) 111;
  }

  printf("%ld\n", (long) p); // CHECK: 111
```

And as I mentioend, even if there is a pointer attachment, and the value of p cannot be modified in the region, a user may still assert that the address of p in the device matches its mapped address, like:

```c
int x;
int *p = &x;
#pragma omp target enter data map(p)
int64_t mapped_p = (int64_t) omp_get_mapped_ptr(&p, omp_get_default_device());

#pragma omp target map(p, p[0:1]) // pointer-attachment succeeds here
{
 assert(mapped_p == (int64_t) &p && "address of p in device should match the mapped address before the region.");
}
```
